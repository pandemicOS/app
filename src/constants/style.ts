export const STYLE = {
  FONT: {
    SIZE: {
      HEADING_BIG: 32,
      HEADING_1: 23,
      HEADING_2: 17,
      BUTTON_LARGE: 17,
      PARAGRAPH: 15,
      PARAGRAPH_SMALL: 13
    },
    WEIGHTS: {
      LIGHT: '300' as any,
      REGULAR: '400' as any,
      MEDIUM: '600' as any,
      BOLD: '700' as any
    }
  },
  COLOR: {
    PRIMARY: '#FFBB00',
    SECONDARY: '#654400',
    TERTIARY: '#7C7C7C',
    QUATERNARY: '#4A90E2',
    QUINARY: '#F6F6F6',
    SENARY: '#646565',
    SEPTENARY: '#1ED66C',
    OCTANARY: '#D5DEF1',
    NONARY: '#74B4FF',
    DENARY: '#3A93FB',
    BLACK: '#1E272E',
    WHITE: '#FFF',
    WHITE_SMOKE: '#FAFAFA',
    LIGHT_GRAY: '#D2D2D2',
    SHADOW_GRAY: '#D7D7D7',
    DIM_GRAY: '#676767',
    DARK_GRAY: '#898989',
    LIGHTEST_GRAY: '#E7E7E7',
    MID_GRAY: '#6A6A77',
    ALMOST_BLACK: '#303030',
    RED_CARDINAL: '#C70D32',
    DARK_RED: '#FF0000',
    CURIOUS_BLUE: '#4191CE',
    INVISIBLE: 'rgba(0, 0, 0, 0)',
    INVISIBLE_WHITE: 'rgba(255, 255, 255, 0)',
    GREEN: '#7ED321',
    RED: '#CE384A',
  },
  BORDER: {
    GLOBAL_RADIUS: 4,
    QUARTER: 25
  }
};
