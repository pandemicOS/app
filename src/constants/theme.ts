import { Dimensions, StyleSheet } from 'react-native';
import { STYLE } from './style';

export const CONTAINER_MAX_WIDTH = 600;
const screenSizeW = Dimensions.get('window').width;
const screenPadding =
  screenSizeW >= CONTAINER_MAX_WIDTH
    ? (screenSizeW - CONTAINER_MAX_WIDTH) / 2
    : 40;

// Generic navigation properties
const navigation = {
  elevation: 0,
  shadowOpacity: 0,
  shadowOffset: {
    height: 0,
    width: 0
  },
  shadowRadius: 0,
  shadowColor: 'transparent',
  borderBottomWidth: 0
};

export const THEME = {
  screenPadding: {
    paddingHorizontal: screenPadding
  },
  navHeader: {
    backgroundColor: STYLE.COLOR.WHITE,
    height: 60,
    borderBottomWidth: 0,
    elevation: 0
  },
  backButton: {
    marginLeft: 25,
    width: 40,
    height: 40,
    justifyContent: 'center'
  },
  menuButton: {
    width: 40,
    height: 40,
    marginLeft: 20,
    justifyContent: 'center'
  },
  navHeaderTitleText: {
    fontSize: 16,
    flex: 1,
    color: STYLE.COLOR.BLACK,
    textAlign: 'center',
    alignSelf: 'center',
  },
  navHeaderGray: {
    ...navigation,
    backgroundColor: STYLE.COLOR.WHITE
  },
  navHeaderWhite: {
    ...navigation,
    backgroundColor: STYLE.COLOR.WHITE
  },
  navHeaderButton: {
    padding: 10
  },
  navLogo: {
    height: 29,
    width: 134,
    marginRight: 20
  },
  regularText: {
    color: STYLE.COLOR.DIM_GRAY,
    fontWeight: STYLE.FONT.WEIGHTS.REGULAR,
  },
  boldText: {
    color: STYLE.COLOR.DIM_GRAY,
    fontWeight: STYLE.FONT.WEIGHTS.BOLD,
  },
  titleText: {
    marginBottom: 20,
    textAlign: 'center',
    fontSize: 20,
    fontWeight: STYLE.FONT.WEIGHTS.BOLD,
    color: STYLE.COLOR.CURIOUS_BLUE
  },
  textError: {
    color: STYLE.COLOR.RED_CARDINAL,
    backgroundColor: 'transparent'
  },
  errorWrapper: {
    marginTop: 8,
    marginBottom: 16
  }
} as any;
