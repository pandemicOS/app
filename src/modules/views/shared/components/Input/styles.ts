import { StyleSheet } from 'react-native';

import { STYLE } from '../../../../../constants';

const styles = StyleSheet.create({
  viewTextInput: {
    alignItems: 'center',
    flexDirection: 'row',
    minHeight: 30,
    borderBottomWidth: 1,
    marginTop: 10,
    backgroundColor: STYLE.COLOR.INVISIBLE
  },
  viewIconTextInput: {
    width: 20,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: STYLE.COLOR.INVISIBLE
  },
  textInput: {
    width: '100%',
    color: STYLE.COLOR.BLACK,
    backgroundColor: STYLE.COLOR.INVISIBLE,
    fontSize: 16,
    minHeight: 4,
    fontWeight: STYLE.FONT.WEIGHTS.MEDIUM
  },
  datePickerPlaceholder: {
    color: STYLE.COLOR.LIGHT_GRAY,
    fontWeight: 'bold'
  },
  textError: {
    textAlign: 'left',
    color: STYLE.COLOR.RED_CARDINAL,
    backgroundColor: STYLE.COLOR.INVISIBLE
  },
  errorWrapper: {
    marginTop: 8,
    marginBottom: 10
  },
  invalidText: {
    borderBottomColor: STYLE.COLOR.DARK_RED,
  },
  validText: {
    borderBottomColor: STYLE.COLOR.LIGHTEST_GRAY,
  }
});

export const customStyles = {
  dateStyle: {
    dateInput: {
      borderWidth: 0,
      alignItems: 'flex-start',
      paddingTop: 12
    },
    placeholderText: styles.datePickerPlaceholder,
    dateText: styles.textInput,
    btnTextConfirm: { color: STYLE.COLOR.CURIOUS_BLUE },
    btnTextCancel: { color: STYLE.COLOR.DARK_RED }
  }
};

export default styles;
