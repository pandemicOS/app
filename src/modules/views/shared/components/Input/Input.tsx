import React from 'react';
import {
  Image,
  View,
  Text,
  NativeSyntheticEvent,
  TextInputSubmitEditingEventData,
  KeyboardTypeOptions,
  StyleProp,
  ViewStyle,
  ReturnKeyTypeOptions,
  TextInput
} from 'react-native';
import { TextInputMask, TextInputMaskTypeProp } from 'react-native-masked-text';

import styles, { customStyles } from './styles';
import { STYLE } from '../../../../../constants';

export interface IInputProps {
  placeholder?: string;
  value?: string;
  onChangeText?: (value: string) => void;
  icon?: number;
  keyboardType?: string;
  secureTextEntry?: boolean;
  error?: string;
  mask?: string;
  maxLength?: number;
  type?: TextInputMaskTypeProp;
  returnKeyType?: ReturnKeyTypeOptions;
  dateTimeFormat?: string;
  isDisabled?: boolean;
  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters';
  multiline?: boolean;
  onValidateShowError?: (text: string) => string;
  onSubmitEditing?: (value?: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => void;
  onEndEditing?: () => void;
  style?: StyleProp<ViewStyle>;
}

export interface IInputState {
  isValid: boolean;
  hasValidated: boolean;
  error: string;
}

export default class Input extends React.PureComponent<IInputProps, IInputState> {
  private static defaultProps = {
    onSubmitEditing: undefined,
    onEndEditing: undefined,
    onChangeText: undefined,
    onValidateShowError: undefined,
    children: null,
    mask: '*'.repeat(100),
    type: 'custom',
    returnKeyType: 'default',
    dateTimeFormat: 'MM/DD/YYYY',
    isDisabled: false,
    autoCapitalize: 'none',
    multiline: false,
  } as IInputProps;

  public state = {
    hasValidated: false,
    error: null
  } as IInputState;

  public input: TextInput;

  public isValid = (): boolean => {
    const { onValidateShowError } = this.props;
    const { error, hasValidated } = this.state;
    return onValidateShowError === undefined || (!error && hasValidated);
  }

  public validate = (text: string) => {
    const { onValidateShowError } = this.props;
    if (onValidateShowError) {
      this.setState({ error: onValidateShowError(text ? text.trim() : ''), hasValidated: true });
    }
  }

  public resetValidation = () => {
    this.setState({ error: null, hasValidated: false });
  }

  public render() {
    const {
      placeholder,
      value,
      onChangeText,
      icon,
      keyboardType,
      secureTextEntry,
      onEndEditing,
      onSubmitEditing,
      children,
      mask,
      maxLength,
      type,
      returnKeyType,
      dateTimeFormat,
      autoCapitalize,
      style,
      isDisabled,
      multiline,
    } = this.props;

    const { error } = this.state;

    return (
      <View>
        <View style={[styles.viewTextInput, error ? styles.invalidText : styles.validText, style, { opacity: isDisabled ? 0.5 : 1 }]}>
          {icon && (
            <View style={styles.viewIconTextInput}>
              <Image source={icon} />
            </View>
          )}
          <TextInputMask
            refInput={input => {
              this.input = input;
            }}
            type={type}
            options={type === 'datetime' ? { format: dateTimeFormat } : { mask }}
            maxLength={maxLength}
            autoCapitalize={autoCapitalize}
            underlineColorAndroid={'rgba(0,0,0,0)'}
            style={[styles.textInput, icon && { paddingLeft: 10 }]}
            placeholder={placeholder}
            placeholderTextColor={STYLE.COLOR.LIGHT_GRAY}
            value={value}
            onChangeText={onChangeText}
            keyboardType={keyboardType as KeyboardTypeOptions}
            returnKeyType={keyboardType === 'numeric' || keyboardType === 'number-pad' ? 'done' : returnKeyType}
            secureTextEntry={secureTextEntry}
            onSubmitEditing={text => onSubmitEditing && onSubmitEditing(text)}
            onEndEditing={(event: any) => {
              this.validate(event.nativeEvent.text);
            }}
            multiline={multiline}
            editable={!isDisabled}
          />
          {children}
        </View>
        {error ? (
          <View style={styles.errorWrapper}>
            <Text style={styles.textError}>
              {error}
            </Text>
          </View>
        ) : null}
      </View>
    );
  }
}
