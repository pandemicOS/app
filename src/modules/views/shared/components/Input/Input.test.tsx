import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Input, { IInputProps } from './Input';
import { IMAGES } from '../../../../../constants';
import { View } from 'react-native';

jest.mock('react-native-datepicker');
jest.mock('react-native-masked-text');

describe('Input', () => {
  let props: IInputProps;
  beforeEach(() => {
    props = {
      placeholder: '',
      value: '',
      onChangeText: jest.fn(),
      icon: IMAGES.ICONS.CALENDAR,
      keyboardType: 'default',
      secureTextEntry: false,
      onSubmitEditing: jest.fn(),
      style: {}
    };
  });
  it('renders without crashing', () => {
    const rendered = renderer.create(<Input {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders with DatePicker without crashing', () => {
    props.keyboardType = 'date';
    const rendered = renderer.create(<Input {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders with DatePicker without icon', () => {
    props.keyboardType = 'date';
    props.icon = null;
    const rendered = renderer.create(<Input {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('renders with error', () => {
    const wrapper = shallow(<Input {...props} />);
    const elementCount = wrapper.find(View).length;
    wrapper.instance().setState({ error: 'error' });
    expect(wrapper.find(View).length).toBeGreaterThan(elementCount);
  });

  it('should call onSubmitEditing', () => {
    const wrapper = shallow(<Input {...props} />);
    (wrapper.find('TextInputMask').at(0).prop('onSubmitEditing') as any)();
    expect(props.onSubmitEditing).toBeCalled();
  });

  it('should call onEndEditing', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow(<Input {...props} />);
    (wrapper.find('TextInputMask').at(0).prop('onEndEditing') as any)({ nativeEvent: { text: 'text' } });
    const newState = wrapper.instance().state;
    expect(newState).toEqual({ error: 'error', hasValidated: true });
  });

  it('should validate with null text', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Input>(<Input {...props} />);
    (wrapper.find('TextInputMask').at(0).prop('onEndEditing') as any)({ nativeEvent: { text: null } });
    const newState = wrapper.instance().state;
    expect(wrapper.instance().props.onValidateShowError).toBeCalled();
  });

  it('should call onEndEditing but not validate', () => {
    const wrapper = shallow(<Input {...props} />);
    (wrapper.find('TextInputMask').at(0).prop('onEndEditing') as any)({ nativeEvent: { text: 'text' } });
    const newState = wrapper.instance().state;
    expect(newState).toEqual({ error: null, hasValidated: false });
  });

  it('should not call onSubmitEditing', () => {
    const wrapper = shallow(<Input {... {...props, onSubmitEditing: undefined}} />);
    (wrapper.find('TextInputMask').at(0).prop('onSubmitEditing') as any)();
    expect(props.onSubmitEditing).toHaveBeenCalledTimes(0);
  });

  it('should change date with onSubmitEditing', () => {
    props.keyboardType = 'date';
    const wrapper = shallow(<Input {...props} />);
    (wrapper.find('DatePicker').at(0).prop('onDateChange') as any)();
    expect(props.onChangeText).toBeCalled();
    expect(props.onSubmitEditing).toBeCalled();
  });

  it('should change date without onSubmitEditing', () => {
    props.keyboardType = 'date';
    const wrapper = shallow(<Input {... {...props, onSubmitEditing: undefined}} />);
    (wrapper.find('DatePicker').at(0).prop('onDateChange') as any)();
    expect(props.onChangeText).toBeCalled();
    expect(props.onSubmitEditing).toHaveBeenCalledTimes(0);
  });

  it('should resetValidation', () => {
    const wrapper = shallow<Input>(<Input {...props} />);
    wrapper.instance().setState({ error: 'error', hasValidated: true });
    wrapper.instance().resetValidation();
    const newState = wrapper.instance().state;
    expect(newState).toEqual({ error: null, hasValidated: false });
  });

  it('isValid with undefined onValidateShowError should return true', () => {
    const wrapper = shallow<Input>(<Input {...props} />);
    expect(wrapper.instance().isValid()).toBe(true);
  });

  it('isValid with error should return false', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Input>(<Input {...props} />);
    wrapper.instance().setState({ error: 'error', hasValidated: true });
    expect(wrapper.instance().isValid()).toBe(false);
  });

  it('isValid without error and previously validated should return true', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Input>(<Input {...props} />);
    wrapper.instance().setState({ error: null, hasValidated: true });
    expect(wrapper.instance().isValid()).toBe(true);
  });

  it('isValid without error but not previously validated should return false', () => {
    props.onValidateShowError = jest.fn().mockReturnValue('error');
    const wrapper = shallow<Input>(<Input {...props} />);
    wrapper.instance().setState({ error: null, hasValidated: false });
    expect(wrapper.instance().isValid()).toBe(false);
  });
});
