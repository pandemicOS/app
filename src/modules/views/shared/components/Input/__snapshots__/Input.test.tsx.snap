// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders with DatePicker without crashing 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "borderBottomWidth": 0.5,
          "flexDirection": "row",
          "height": 42,
          "marginTop": 10,
        },
        Object {
          "borderBottomColor": "#E7E7E7",
        },
        Object {},
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "justifyContent": "center",
          "width": 20,
        }
      }
    >
      <Image
        source={1}
      />
    </View>
  </View>
</View>
`;

exports[`Input renders with DatePicker without icon 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "borderBottomWidth": 0.5,
          "flexDirection": "row",
          "height": 42,
          "marginTop": 10,
        },
        Object {
          "borderBottomColor": "#E7E7E7",
        },
        Object {},
        Object {
          "opacity": 1,
        },
      ]
    }
  />
</View>
`;

exports[`Input renders without crashing 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "borderBottomWidth": 0.5,
          "flexDirection": "row",
          "height": 42,
          "marginTop": 10,
        },
        Object {
          "borderBottomColor": "#E7E7E7",
        },
        Object {},
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "justifyContent": "center",
          "width": 20,
        }
      }
    >
      <Image
        source={1}
      />
    </View>
  </View>
</View>
`;
